.github/workflows/tests.yml

```diff
@@ -0,0 +1,74 @@
+name: Run all tests
+
+# Run this workflow every time a new commit is pushed to the repository
+on:
+  push:
+    branches:
+    - master
+  pull_request:
+
+jobs:
+  run-all-tests:
+    name: Run all tests
+
+    # TODO: not sure what the best OS to run on is so ubuntu will do for now
+    runs-on: ubuntu-latest
+
+    strategy:
+      # Allows running the build multiple times with different configurations
+      matrix:
+        rust: ["stable", "beta"]
+        deny_warnings: ['', 1]
+
+        # TODO: I don't know how to make it run only the check_exercises.sh script
+        # For nightly rust, run the job with DENYWARNINGS=1
+        include:
+          - rust: nightly
+            deny_warnings: 1
+    
+    # Allow failures on nightly rust, or beta if DENYWARNINGS is enabled
+    continue-on-error: ${{ matrix.rust }} == "nightly" || ( ${{ matrix.rust }} == "beta" && DENYWARNINGS == 1 )
+
+    steps:
+      # Checks out a copy of your repository on the ubuntu-latest machine
+      - name: Checkout code
+        uses: actions/checkout@v2
+        env: 
+          DENYWARNINGS: ${{ matrix.deny_warnings }}
+
+      # run these as steps
+      - name: Check exercises
+        run: ./_test/check-exercises.sh
+
+      - name: Ensure src/lib.rs files exist
+        run: bash ./_test/ensure-lib-src-rs-exist.sh
+
+      - name: Ensure stubs compile
+        run: sh ./_test/ensure-stubs-compile.sh
+
+      - name: Count ignores
+        run: sh ./_test/count-ignores.sh
+
+      - name: Fetch configlet
+        run: ./bin/fetch-configlet
+
+      - name: Check configlet format
+        run: ./_test/check-configlet-fmt.sh
```

Comment:
There are a few sequential dependencies in here; this one, in particular, requires configlet from the previous "Fetch configlet" step. The [`needs` key](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds) seems designed for this situation.

Reply:
You only need the `needs` key if you have multiple jobs in a workflow. Steps only run if the previous step succeeded by default. This workflow only has one job with many steps. 

(Splitting them up to different jobs maybe be useful though, for example to run configlet independant of exercise tests.)

Reply:
I see. In that case, we should probably split this up into a few different jobs which can run in parallel. Something like:

- **compilation job**: ensure stubs compile, check exercises, check-exercise-crate
- **ensuring conventions are followed job**: ensure src/lib.rs files exist, count ignores, check uuids, verify exercise difficulties, check exercises for authors, 
- **configlet job**: fetch configlet, check configlet format, configlet lint, ensure readmes are updated

One nice thing about splitting out the jobs like that is that only the compilation job actually needs to run on a matrix.